### PR TITLE
fix(wasi-nn): replace hardcoded output sizes with dynamic max_size in RPC

### PR DESCRIPTION
**Pull request by Manar-Elhabbal7**
## Description

The WASI-NN RPC implementation previously used a hardcoded 64 KiB memory size
for `get_output` and `get_output_single`, which silently truncated any output
larger than one Wasm page and made the buffer size impossible to control from
the caller.

This PR replaces the hardcoded value with a caller-supplied `max_size` field in
the gRPC protocol, and adds bounds validation and null-pointer checks on the
server side.

## Key Changes

- **Protocol** (`wasi_ephemeral_nn.proto`): added `max_size` field to
  `GetOutputRequest` so the client declares how large an output buffer it needs.
- **Server** (`wasi_nn_rpcserver.h`):
  - `HostFuncCaller` now converts the requested byte count to Wasm pages
    (64 KiB each) dynamically instead of hardcoding 1 page.
  - Added input validation: rejects `max_size == 0` and enforces a 16 MiB cap
    to prevent a malicious client from crashing the server with a huge
    allocation request.
  - Added integer-overflow guard before computing `BufPtr + MaxSize`.
  - Added null-pointer checks on `getPointer` results before dereferencing.
  - Added `spdlog::debug` tracing so allocation decisions are visible at
    debug log level.
- **Client** (`wasinnfunc.cpp`): populates `max_size` from the actual
  `OutBufferMaxSize` the Wasm guest requested.
- **Tests** (`test/plugins/wasi_nn/wasi_nn.cpp`): added boundary cases for
  both `get_output` and `get_output_single` — zero, exact page boundary,
  multi-page, 16 MiB exact, 16 MiB + 1, `UINT32_MAX`, memory-constrained
  guest, and exact-fit scenarios.

## Checklist

- [x] **DCO Signed-off**: All commits are signed-off (`git commit -s`).
- [x] **Commit Messages**: Verified with `commitlint` against Conventional Commits.
- [x] **Local Tests Passed**: `ctest -j$(nproc) --output-on-failure` — 33/33 passed.

## Test Evidence

Custom RPC test with variable tensor sizes confirming correct Wasm page
allocation:

RPC tensor size test

Full `ctest` run — 33/33 passed:

ctest results

### CODE REVIEW

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h [hunk 1]
```diff
@@ -51,13 +51,16 @@ void writeBinaries(WasmEdge::Runtime::Instance::MemoryInstance &MemInst,
 class HostFuncCaller {
 public:
   HostFuncCaller(const Runtime::Instance::ModuleInstance &NNM,
-                 std::string_view FuncName, uint32_t MemorySize) noexcept
+                 std::string_view FuncName, uint32_t MemorySizeInBytes) noexcept
       : FuncInst(NNM.findFuncExports(FuncName)), FuncName(FuncName), Mod(""),
         Frame(nullptr, &Mod) {
+    uint32_t MemorySizeInPages = (MemorySizeInBytes + 65535) / 65536;
+    spdlog::info("[WASI-NN-RPC] Function: {}, Request Bytes: {}, Allocated Pages: {}",
```
Comment (hydai) [inline include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h]: You are not following the same logging style we used in this header.

**Review by hydai — CHANGES_REQUESTED** [review]

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h (hunk 1 above)
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h]: i edit it now

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h [hunk 2]
```diff
@@ -319,10 +322,11 @@ class GraphExecutionContextResourceService final
     std::string_view FuncName = "get_output"sv;
     uint32_t ResourceHandle = RPCRequest->resource_handle();
     uint32_t Index = RPCRequest->index();
-    uint32_t MemorySize = UINT32_C(65536); // FIXME
+    uint32_t MaxSize = RPCRequest->max_size();
     uint32_t BytesWrittenPtr = UINT32_C(0);
     uint32_t BufPtr = BytesWrittenPtr + UINT32_C(4);
-    uint32_t BufMaxSize = MemorySize - BufPtr;
+    uint32_t BufMaxSize = MaxSize;
+    uint32_t MemorySize = BufPtr + BufMaxSize;
```
[the remark this answers is not in the document]
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h]: done

**Comment by Manar-Elhabbal7** [comment]
@hydai i resolved the copilot reviews and the log are now match with the standard log for this header

**Comment by hydai** [comment]
You should resolve the copilot threads if you have already fixed them.

**Comment by Manar-Elhabbal7** [comment]
> You should resolve the copilot threads if you have already fixed them.

done . resolved all of them

**Comment by Manar-Elhabbal7** [comment]
> ## Pull request overview
> Copilot reviewed 3 out of 3 changed files in this pull request and generated 5 comments.

hello @hydai and  @dm4   i also resolved these issues
i wish it better now :) . You can review my pr

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h [hunk 3]
```diff
@@ -319,10 +328,24 @@ class GraphExecutionContextResourceService final
...
     uint32_t BytesWrittenPtr = UINT32_C(0);
     uint32_t BufPtr = BytesWrittenPtr + UINT32_C(4);
-    uint32_t BufMaxSize = MemorySize - BufPtr;
+    uint32_t BufMaxSize = MaxSize;
+
+    static constexpr uint32_t MaxOutputBytesLimit = 16U * 1024U * 1024U;
+    if (MaxSize == 0 || MaxSize > MaxOutputBytesLimit) {
+      RPCContext->AddTrailingMetadata("errno", "1");
+      return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
+                          fmt::format("max_size must be in 1..{} bytes"sv,
```
[the remark this answers is not in the document]
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h]: i've clarified in  the PR description that i added a 16 MiB output limit to prevent attackers from crashing the server with huge requests.

**Comment by Manar-Elhabbal7** [comment]
@hydai I've resolved the Copilot suggestions, but I'm finding that 
some of them don't always work well or cause issues later. so if u have particular changes u can mention them .

**Comment by hydai** [comment]
Hi @Manar-Elhabbal7,  
That's possible. Since it's an AI review from Copilot, you can resolve those nonsensical review comments.

**Review by hydai — CHANGES_REQUESTED** [review]
Could you add a test to cover this situation?

**Comment by Manar-Elhabbal7** [comment]
> ## Pull request overview
> Copilot reviewed 4 out of 4 changed files in this pull request and generated 4 comments.

I have resolved all of them and refactored the pr title and description. 

Also I refactored the code to use helper function to avoid duplication and removed non necessary code

If it's now good I will make squash to make the commits clearner

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h [hunk 4]
```diff
@@ -4,7 +4,11 @@
 #include "runtime/instance/module.h"
 #include "wasi_ephemeral_nn.grpc.pb.h"
 
+#include "common/spdlog.h"
```
Comment (hydai) [inline include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h]: This line should be placed above.

**Review by hydai — CHANGES_REQUESTED** [review]

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h (hunk 4 above)
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h]: okay . done

**Comment by hydai** [comment]
Please rebase your commits onto the latest master branch to resolve the conflicts.

**Comment by Manar-Elhabbal7** [comment]
> Please rebase your commits onto the latest master branch to resolve the conflicts.

done . I also run all the tests again and all of them passed successfully 

image

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h [hunk 5]
```diff
@@ -84,8 +87,11 @@ class HostFuncCaller {
                  std::string_view FuncName, uint64_t MemoryBytes) noexcept
       : FuncInst(NNM.findFuncExports(FuncName)), FuncName(FuncName), Mod(""),
         Frame(nullptr, &Mod) {
-    const uint32_t Pages = static_cast<uint32_t>(
-        (MemoryBytes + UINT64_C(65535)) / UINT64_C(65536));
+    const uint64_t PageSize = Runtime::Instance::MemoryInstance::kPageSize;
+    const uint64_t Pages = (MemoryBytes + PageSize - 1) / PageSize;
+    spdlog::debug(
+        "[WASI-NN-RPCSERVER] HostFuncCaller: function={}, bytes={}, pages={}"sv,
+        FuncName, MemoryBytes, Pages);
```
Comment (hydai) [inline include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:94]: Does this information require notifying users when they are in debug mode? If so, what is the reason that other places do not notify users?

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h [hunk 6]
```diff
@@ -378,31 +405,67 @@ class GraphExecutionContextResourceService final
...
+    if (BytesWrittenInMem == nullptr) {
+      RPCContext->AddTrailingMetadata("errno", "5");
+      return grpc::Status(grpc::StatusCode::RESOURCE_EXHAUSTED,
+                          "failed to allocate output buffer for BytesWritten"s);
+    }
+    auto BytesWritten = *BytesWrittenInMem;
+    if (BytesWritten > BufMaxSize) {
+      spdlog::error(
+          "[WASI-NN-RPCSERVER] {}: host function wrote more bytes ({}) than BufMaxSize ({})"sv,
+          FuncName, BytesWritten, BufMaxSize);
```
Comment (hydai) [inline include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:460]: I found that some of the errors notify users with spdlog, but others don't. Would you like to explain the reasoning behind your choices?

**Review by hydai — CHANGES_REQUESTED** [review]

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h (hunk 5 above)
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:94]: The reason is that   `HostFuncCaller ` constructor is shared and reused across all handlers (like  Load , 
LoadByName , etc.). so it used in  all host calls. no need to write it again 
i have a question spdlog  is meant for developers/operators debugging the runtime server, not to notify  users right?

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h (hunk 6 above)
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:460]: I used  `spdlog::error ` here because the host function writing more bytes than  max_size  represents a      
  serious  bug . For other errors like validation ones,i handeled  by returning the gRPC status code to the  client

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h (hunk 5 above)
Comment (hydai) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:94]: Actually, it's for users to report issues to the developers. So, my question is: which parts do you want to show to users and developers? It seems like there is no specific rule in your PR at the moment.

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h (hunk 6 above)
Comment (hydai) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:460]: Ditto. Please ensure that you have a rule for notifying errors.

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h (hunk 5 above)
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:94]: Hello @hydai  My rule is simple I use `spdlog::error` for internal/server-side errors that developers need to be notified about, and I use the status code for everything else (i.e. expected outcomes of bad client input).
I had already applied this rule, but I missed the two cases you pointed out  thanks a lot for addressing them
and feel free to give another feedbacks

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h (hunk 6 above)
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:460]: > Ditto. Please ensure that you have a rule for notifying errors.

yeah , i explained it in the prev requested changee

**Review by hydai — CHANGES_REQUESTED** [review]
The RPC CI failed.

**Comment by Manar-Elhabbal7** [comment]
@hydai PTAL
I don't know why it was failing maybe it was a flaky test i raised in my fork a pr to see the actions here 
some tests failed but i think they are not related to my code changes

https://github.com/Manar-Elhabbal7/WasmEdge/pull/2

**Comment by Manar-Elhabbal7** [comment]
Hello @hydai , PTAL

i raised a pr in my fork to show the ci actions and RPC ci now passed

https://github.com/Manar-Elhabbal7/WasmEdge/pull/2#

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h [hunk 7]
```diff
@@ -84,8 +87,11 @@ class HostFuncCaller {
                  std::string_view FuncName, uint64_t MemoryBytes) noexcept
       : FuncInst(NNM.findFuncExports(FuncName)), FuncName(FuncName), Mod(""),
         Frame(nullptr, &Mod) {
-    const uint32_t Pages = static_cast<uint32_t>(
-        (MemoryBytes + UINT64_C(65535)) / UINT64_C(65536));
+    const uint64_t PageSize = Runtime::Instance::MemoryInstance::kPageSize;
+    const uint64_t Pages = (MemoryBytes + PageSize - 1) / PageSize;
```
[the remark this answers is not in the document]
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:91]: `MemoryType(Pages)`  is uint64_t already not uint32_t

File: include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h [hunk 8]
```diff
@@ -378,31 +405,78 @@ class GraphExecutionContextResourceService final
...
+        if (Needed != nullptr) {
+          RPCContext->AddTrailingMetadata("required_size",
+                                          std::to_string(*Needed));
+          return grpc::Status(
+              grpc::StatusCode::RESOURCE_EXHAUSTED,
+              fmt::format("output buffer too small, need {} bytes"sv, *Needed));
+        }
+        return grpc::Status(grpc::StatusCode::RESOURCE_EXHAUSTED,
+                            "output buffer too small"s);
       }
```
[the remark this answers is not in the document]
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:441]: @hydai, can you help me with this ? I suggest keeping it as it is without retry

i have  also made the other copilot suggestions and refactoring as they were make sense for me

[the remark this answers is not in the document]
Comment (hydai) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:441]: Could you please explain why you don't want to add the retry?

[the remark this answers is not in the document]
Comment (Manar-Elhabbal7) [reply include/driver/wasi_nn_rpc/wasi_nn_rpcserver/wasi_nn_rpcserver.h:441]: > Could you please explain why you don't want to add the retry?


Hi @hydai  , i spent time to see what's pros and cons of each choice and i ended up with not to allow retry for the following reason :

1. first things first : adding the retry will not solve the problem  even if the server retries and allocates a larger buffer  the wasm guest still has the original smaller buffer in the linear memory. so that the cilent will still fail to write the data and this waste the memory and compute in the process 

2. and also this alow for automatic retries and let clients send requrest on purpose to force the server into repeated large allocations  . my approach now to return `ErrNo::TooLarge` along with `required_size` in the gRPC metadata i think it much better the server just reporst the exactly how many btyes needed and the guest have the full control of its own memory allocation

which was here in  lines `wasi_nn_rpcserver.h:428-441`

Do u have any other suggestions ?

**Comment by hydai** [comment]
Squash your commits into one and rebase onto the latest master branch.

**Comment by Manar-Elhabbal7** [comment]
> Squash your commits into one and rebase onto the latest master branch.

done : )

**Review by hydai — APPROVED** [review]
